packages/slate/src/controllers/editor.js

```diff
@@ -0,0 +1,312 @@
+import Debug from 'debug'
+
+import AbstractChange from './change'
+import CorePlugin from '../plugins/core'
+import CommandsPlugin from '../plugins/commands'
+import QueriesPlugin from '../plugins/queries'
+import SchemaPlugin from '../plugins/schema'
+import Value from '../models/value'
+
+/**
+ * Debug.
+ *
+ * @type {Function}
+ */
+
+const debug = Debug('slate:editor')
+
+/**
+ * Editor.
+ *
+ * @type {Editor}
+ */
+
+class Editor {
+  /**
+   * Create a new `Editor` with `attrs`.
+   *
+   * @param {Object} attrs
+   * @param {Object} options
+   */
+
+  constructor(attrs = {}, options = {}) {
+    const { editor = this } = options
+    const {
+      onChange = () => {},
+      plugins = [],
+      readOnly = false,
+      value = Value.create(),
+    } = attrs
+
+    this.editor = editor
+
+    this.tmp = {
+      change: null,
+      isChanging: false,
+      lastPlugins: null,
+      lastValue: null,
+      rawPlugins: null,
+    }
+
+    this.setProperties({ onChange, plugins, readOnly, value }, options)
+  }
+
+  /**
+   * Perform a change on the editor, passing `...args` to `change.call`.
+   *
+   * @param {Mixed} ...args
+   */
+
+  change = (...args) => {
+    const { Change, editor, value } = this
+    const { isChanging } = this.tmp
+    const change = isChanging ? this.tmp.change : new Change({ value, editor })
+
+    try {
+      this.tmp.change = change
+      this.tmp.isChanging = true
+      change.call(...args)
+    } catch (error) {
+      throw error
+    } finally {
+      this.tmp.isChanging = isChanging
+    }
+
+    // If this isn't the top-most change function, exit to let it finish.
+    if (isChanging === true) {
+      return
+    }
+
+    this.run('onChange', change)
+
+    // If the change doesn't define any operations to apply, abort.
+    if (change.operations.size === 0) {
+      return
+    }
+
+    // Store a reference to the last `value` and `plugins` that were seen by the
+    // editor, so we can know whether to normalize a new unknown value if one
+    // is passed in via `this.props`.
+    this.tmp.lastValue = change.value
```

Comment:
Is `this.tmp.lastValue` always normalized? If so, could we call this `lastNormalizedValue`?